src/common/tusb_fifo.h

```diff
@@ -86,37 +99,30 @@ static inline void tu_fifo_config_mutex(tu_fifo_t *f, tu_fifo_mutex_t mutex_hdl)
 }
 #endif
 
-bool     tu_fifo_write   (tu_fifo_t* f, void const * p_data);
-uint16_t tu_fifo_write_n (tu_fifo_t* f, void const * p_data, uint16_t count);
+bool     tu_fifo_write                  (tu_fifo_t* f, void const * p_data);
+uint16_t tu_fifo_write_n                (tu_fifo_t* f, void const * p_data, uint16_t count);
 
-bool     tu_fifo_read    (tu_fifo_t* f, void * p_buffer);
-uint16_t tu_fifo_read_n  (tu_fifo_t* f, void * p_buffer, uint16_t count);
+bool     tu_fifo_read                   (tu_fifo_t* f, void * p_buffer);
+uint16_t tu_fifo_read_n                 (tu_fifo_t* f, void * p_buffer, uint16_t count);
 
-bool     tu_fifo_peek_at (tu_fifo_t* f, uint16_t pos, void * p_buffer);
+bool     tu_fifo_peek_at                (tu_fifo_t* f, uint16_t pos, void * p_buffer);
+uint16_t tu_fifo_peek_at_n              (tu_fifo_t* f, uint16_t pos, void * p_buffer, uint16_t n);
 
-static inline bool tu_fifo_peek(tu_fifo_t* f, void * p_buffer)
-{
-  return tu_fifo_peek_at(f, 0, p_buffer);
-}
+uint16_t tu_fifo_count                  (tu_fifo_t* f);
+bool     tu_fifo_empty                  (tu_fifo_t* f);
+bool     tu_fifo_full                   (tu_fifo_t* f);
+uint16_t tu_fifo_remaining              (tu_fifo_t* f);
+bool     tu_fifo_overflow               (tu_fifo_t* f);
+void     tu_fifo_correct_read_pointer   (tu_fifo_t* f);
 
-static inline bool tu_fifo_empty(tu_fifo_t* f)
-{
-  return (f->count == 0);
-}
+// Pointer modifications intended to be used in combinations with DMAs.
+// USE WITH CARE - NO SAFTY CHECKS CONDUCTED HERE! NOT MUTEX PROTECTED!
+void     tu_fifo_advance_write_pointer  (tu_fifo_t *f, uint16_t n);
```

Comment:
Uff honestly i am lacking time dramatically :D I also need to review the UAC2 PR. I did this DMA stuff before in my [RERO](https://gitlab.com/PaniR/rero_stm32) project (in combination with USART) for an STM32F3 and F4. So i know this works at least for STMs. However, since DMAs are rather specific to the used MCUs i am not sure if it is possible to set up a general test?

Reply:
I will do the test then, test is run on x86, it actually doesn't run any DMA, just plain copying then move pointer forward, then read back data for verification. Test is all faking behavior :)

Reply:
Thank you! Maybe it is even better than somebody else tests the code. If there are any issues or questions don't hesitate to ask! :)